lib/AST/IRToASTVisitor.cpp

```diff
@@ -509,6 +534,24 @@ void IRToASTVisitor::visitGetElementPtrInst(llvm::GetElementPtrInst &inst) {
   auto indexed_type{inst.getPointerOperandType()};
   auto base{GetOperandExpr(inst.getPointerOperand())};
 
+  if (auto arg = llvm::dyn_cast<llvm::Argument>(inst.getPointerOperand())) {
```

Comment:
This seems to handle the GEP case, but what about `ptrtoint`, `bitcast`, tail-calls, etc? What about an alternative approach where we pretend that arguments are like `phi` nodes and thus need to be "solved" via `alloca`s. Then, the `byval` arguments become the normal case, and the non-byval ones are immediately put into `alloca`s. Or is this crazy? Mostly, I'm worried about losing track of the byval pointer somehow.

Reply:
Here's a new test case for you: https://godbolt.org/z/6fz8hbzd9

```c++
struct big_base {
  int x[32];
};
struct big_derived : public big_base {

};
extern double bar(struct big_base y);
double foo(struct big_derived x) {
  return bar(x);
}
```

Which produces the following:

```llvm
; Function Attrs: mustprogress uwtable
define dso_local double @_Z3foo11big_derived(%struct.big_derived* nocapture readonly byval(%struct.big_derived) align 8 %0) local_unnamed_addr #0 {
  %2 = getelementptr inbounds %struct.big_derived, %struct.big_derived* %0, i64 0, i32 0
  %3 = tail call double @_Z3bar8big_base(%struct.big_base* nonnull byval(%struct.big_base) align 8 %2)
  ret double %3
}
```

Reply:
And another: https://godbolt.org/z/79E9Mc1dW

```c++
struct big_base {
  int x[32];
};
extern double bar(struct big_base y);
double foo(struct big_base x) {
  return bar(x);
}
```

This one has no GEP:
```llvm
; Function Attrs: mustprogress uwtable
define dso_local double @_Z3foo8big_base(%struct.big_base* nocapture readonly byval(%struct.big_base) align 8 %0) local_unnamed_addr #0 {
  %2 = tail call double @_Z3bar8big_base(%struct.big_base* nonnull byval(%struct.big_base) align 8 %0)
  ret double %2
}
```

Reply:
I've switched to using a more 'general' approach in generating the auxiliary pointer value, this should be robust enough to work for other instructions as well.

I've added a specific test type that decompiles bitcode directly, another approach could be to instead test roundtripping with full optimization too instead.